File: src/lib/services/verification.ts
import { supabase } from '@/lib/supabase';
import { API_URL } from '@/lib/config';

interface VerificationRequest {
  contractAddress: string;
  sourceCode: string;
  contractName: string;
  compilerVersion: string;
  chainId: number;
  projectPath?: string;
}

interface VerificationResult {
  status: 'pending' | 'verified' | 'failed';
  guid?: string;
  message: string;
}

// Get the appropriate Arbiscan API URL and key based on chain ID
function getArbiscanConfig(chainId: number) {
  switch (chainId) {
    case 42161: // Arbitrum One
      return {
        apiUrl: 'https://api.arbiscan.io/v2/api',
        apiKey: import.meta.env.VITE_ARBISCAN_API_KEY,
        explorerUrl: 'https://arbiscan.io',
        chainId: 42161,
      };
    case 421614: // Arbitrum Sepolia
      return {
        apiUrl: 'https://api.arbiscan.io/v2/api',
        apiKey: import.meta.env.VITE_ARBISCAN_SEPOLIA_API_KEY,
        explorerUrl: 'https://sepolia.arbiscan.io',
        chainId: 421614,
      };
    default:
      return null;
  }
}

// Submit contract for verification via backend API
export async function verifyContract(request: VerificationRequest): Promise<VerificationResult> {
  try {
    const response = await fetch(`${API_URL}/api/verification/verify`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        contract_address: request.contractAddress,
        source_code: request.sourceCode,
        contract_name: request.contractName,
        compiler_version: request.compilerVersion,
        chain_id: request.chainId,
      }),
    });

    const data = await response.json();

    if (data.success && data.data) {
      const result = data.data;
      
      // If we got a GUID, wait and check status
      if (result.guid && result.status === 'pending') {
        // Wait a bit before checking status
        await new Promise(resolve => setTimeout(resolve, 3000));
        
        // Check verification status
        const statusResult = await checkVerificationStatus(result.guid, request.chainId);
        
        return {
          status: statusResult.verified ? 'verified' : 'pending',
          guid: result.guid,
          message: statusResult.message || 'Verification submitted successfully',
        };
      }
      
      return {
        status: result.status,
        guid: result.guid,
        message: result.message,
      };
    } else {
      return {
        status: 'failed',
        message: data.message || 'Verification failed',
      };
    }
  } catch (error) {
    console.error('Error verifying contract:', error);
    return {
      status: 'failed',
      message: error instanceof Error ? error.message : 'Failed to verify contract',
    };
  }
}

// Check the status of a verification request via backend API
export async function checkVerificationStatus(guid: string, chainId: number): Promise<{ verified: boolean; message?: string }> {
  try {
    const response = await fetch(`${API_URL}/api/verification/status`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        guid: guid,
        chain_id: chainId,
      }),
    });

    const data = await response.json();

    if (data.success && data.data) {
      const result = data.data;
      return {
        verified: result.status === 'verified',
        message: result.message,
      };
    } else {
      return {
        verified: false,
        message: data.message || 'Failed to check verification status',
      };
    }
  } catch (error) {
    console.error('Error checking verification status:', error);
    return { verified: false, message: 'Failed to check verification status' };
  }
}

// Attempt to verify a Stylus contract specifically
export async function verifyStylusContract(
  contractAddress: string,
  projectId: string,
  chainId: number,
  sourceCode: string,
  contractName: string = 'StylusContract'
): Promise<VerificationResult> {
  const config = getArbiscanConfig(chainId);
  
  if (!config || !config.apiKey || config.apiKey.includes('your-arbiscan')) {
    // No API key configured - mark as failed
    const result: VerificationResult = {
      status: 'failed',
      message: 'Arbiscan API key not configured. Add VITE_ARBISCAN_API_KEY and VITE_ARBISCAN_SEPOLIA_API_KEY to your .env file.',
    };
    
    await updateDeploymentVerificationStatus(contractAddress, chainId, result);
    return result;
  }

  // Attempt verification using the standard Solidity verification process
  // This should work for Stylus contracts on Arbiscan with the v2 API
  const verificationRequest: VerificationRequest = {
    contractAddress,
    sourceCode,
    contractName,
    compilerVersion: 'v0.8.25+commit.b61c2a91', // Use a standard Solidity compiler version
    chainId,
  };

  const result = await verifyContract(verificationRequest);
  
  await updateDeploymentVerificationStatus(contractAddress, chainId, result);
  return result;
}

// Helper function to update deployment verification status
async function updateDeploymentVerificationStatus(
  contractAddress: string,
  chainId: number,
  result: VerificationResult
): Promise<void> {
  try {
    const { error } = await supabase
      .from('deployments')
      .update({
        verification_status: result.status,
        verification_guid: result.guid || null,
        verified_at: result.status === 'verified' ? new Date().toISOString() : null,
      })
      .eq('contract_address', contractAddress)
      .eq('chain_id', chainId);
      
    if (error) {
      console.error('Error updating deployment verification status:', error);
    }
  } catch (error) {
    console.error('Error updating deployment:', error);
  }
}

// Get verification URL for a contract
export function getVerificationUrl(contractAddress: string, chainId: number): string | null {
  const config = getArbiscanConfig(chainId);
  if (!config) return null;
  
  return `${config.explorerUrl}/address/${contractAddress}#code`;
}

File: backend/src/api/verification.rs
use actix_web::{web, HttpResponse};
use serde::{Deserialize, Serialize};
use reqwest;
use std::collections::HashMap;

use super::compile::{ApiResponse, ApiError};

#[derive(Debug, Deserialize)]
pub struct VerificationRequest {
    pub contract_address: String,
    pub source_code: String,
    pub contract_name: String,
    pub compiler_version: String,
    pub chain_id: u64,
}

#[derive(Debug, Deserialize)]
pub struct VerificationStatusRequest {
    pub guid: String,
    pub chain_id: u64,
}

#[derive(Debug, Serialize)]
pub struct VerificationResult {
    pub status: String, // 'pending', 'verified', 'failed'
    pub guid: Option<String>,
    pub message: String,
}

// Get Arbiscan API configuration based on chain ID
fn get_arbiscan_config(chain_id: u64) -> Option<(String, String)> {
    match chain_id {
        42161 => {
            // Arbitrum One
            let api_key = std::env::var("ARBISCAN_API_KEY").ok()?;
            Some(("https://api.arbiscan.io/v2/api".to_string(), api_key))
        }
        421614 => {
            // Arbitrum Sepolia  
            let api_key = std::env::var("ARBISCAN_SEPOLIA_API_KEY")
                .or_else(|_| std::env::var("ARBISCAN_API_KEY")).ok()?;
            Some(("https://api.arbiscan.io/v2/api".to_string(), api_key))
        }
        _ => None,
    }
}

pub async fn verify_contract(req: web::Json<VerificationRequest>) -> HttpResponse {
    let config = match get_arbiscan_config(req.chain_id) {
        Some(config) => config,
        None => {
            return HttpResponse::BadRequest().json(ApiResponse::<()> {
                success: false,
                message: "Verification not supported for this chain".to_string(),
                data: None,
                error: Some(ApiError {
                    code: "UNSUPPORTED_CHAIN".to_string(),
                    message: "Unsupported chain ID".to_string(),
                    details: None,
                }),
            });
        }
    };

    let (api_url, api_key) = config;

    // Prepare form data for Arbiscan API
    let mut form_data = HashMap::new();
    form_data.insert("chainid", req.chain_id.to_string());
    form_data.insert("apikey", api_key);
    form_data.insert("module", "contract".to_string());
    form_data.insert("action", "verifysourcecode".to_string());
    form_data.insert("contractaddress", req.contract_address.clone());
    form_data.insert("sourceCode", req.source_code.clone());
    form_data.insert("codeformat", "solidity-single-file".to_string());
    form_data.insert("contractname", req.contract_name.clone());
    form_data.insert("compilerversion", req.compiler_version.clone());
    form_data.insert("optimizationUsed", "1".to_string());
    form_data.insert("runs", "200".to_string());
    form_data.insert("evmversion", "default".to_string());
    form_data.insert("licenseType", "3".to_string()); // MIT license

    // Make request to Arbiscan API
    let client = reqwest::Client::new();
    let response = match client.post(&api_url).form(&form_data).send().await {
        Ok(response) => response,
        Err(e) => {
            return HttpResponse::InternalServerError().json(ApiResponse::<()> {
                success: false,
                message: "Failed to submit verification request".to_string(),
                data: None,
                error: Some(ApiError {
                    code: "REQUEST_FAILED".to_string(),
                    message: e.to_string(),
                    details: None,
                }),
            });
        }
    };

    let response_body: serde_json::Value = match response.json().await {
        Ok(body) => body,
        Err(e) => {
            return HttpResponse::InternalServerError().json(ApiResponse::<()> {
                success: false,
                message: "Failed to parse API response".to_string(),
                data: None,
                error: Some(ApiError {
                    code: "PARSE_FAILED".to_string(),
                    message: e.to_string(),
                    details: None,
                }),
            });
        }
    };

    // Parse Arbiscan response
    if let Some(status) = response_body.get("status").and_then(|s| s.as_str()) {
        if status == "1" {
            if let Some(guid) = response_body.get("result").and_then(|r| r.as_str()) {
                let result = VerificationResult {
                    status: "pending".to_string(),
                    guid: Some(guid.to_string()),
                    message: "Verification submitted successfully".to_string(),
                };
                
                return HttpResponse::Ok().json(ApiResponse {
                    success: true,
                    message: "Verification submitted".to_string(),
                    data: Some(result),
                    error: None,
                });
            }
        }
    }

    // If we get here, verification failed
    let error_message = response_body
        .get("result")
        .and_then(|r| r.as_str())
        .unwrap_or("Verification failed");

    let result = VerificationResult {
        status: "failed".to_string(),
        guid: None,
        message: error_message.to_string(),
    };

    HttpResponse::Ok().json(ApiResponse {
        success: true,
        message: "Verification failed".to_string(),
        data: Some(result),
        error: None,
    })
}

pub async fn check_verification_status(req: web::Json<VerificationStatusRequest>) -> HttpResponse {
    let config = match get_arbiscan_config(req.chain_id) {
        Some(config) => config,
        None => {
            return HttpResponse::BadRequest().json(ApiResponse::<()> {
                success: false,
                message: "Chain not supported".to_string(),
                data: None,
                error: Some(ApiError {
                    code: "UNSUPPORTED_CHAIN".to_string(),
                    message: "Unsupported chain ID".to_string(),
                    details: None,
                }),
            });
        }
    };

    let (api_url, api_key) = config;

    // Prepare query parameters
    let query_params = [
        ("chainid", req.chain_id.to_string()),
        ("apikey", api_key),
        ("module", "contract".to_string()),
        ("action", "checkverifystatus".to_string()),
        ("guid", req.guid.clone()),
    ];

    let client = reqwest::Client::new();
    let response = match client.get(&api_url).query(&query_params).send().await {
        Ok(response) => response,
        Err(e) => {
            return HttpResponse::InternalServerError().json(ApiResponse::<()> {
                success: false,
                message: "Failed to check verification status".to_string(),
                data: None,
                error: Some(ApiError {
                    code: "REQUEST_FAILED".to_string(),
                    message: e.to_string(),
                    details: None,
                }),
            });
        }
    };

    let response_body: serde_json::Value = match response.json().await {
        Ok(body) => body,
        Err(e) => {
            return HttpResponse::InternalServerError().json(ApiResponse::<()> {
                success: false,
                message: "Failed to parse API response".to_string(),
                data: None,
                error: Some(ApiError {
                    code: "PARSE_FAILED".to_string(),
                    message: e.to_string(),
                    details: None,
                }),
            });
        }
    };

    // Parse status response
    if let Some(status) = response_body.get("status").and_then(|s| s.as_str()) {
        let (verification_status, message) = if status == "1" {
            ("verified".to_string(), "Contract verified successfully".to_string())
        } else {
            let result_msg = response_body
                .get("result")
                .and_then(|r| r.as_str())
                .unwrap_or("Verification status unknown");
            
            if result_msg.contains("Pending") {
                ("pending".to_string(), "Verification is still pending".to_string())
            } else {
                ("failed".to_string(), result_msg.to_string())
            }
        };

        let result = VerificationResult {
            status: verification_status,
            guid: Some(req.guid.clone()),
            message,
        };

        return HttpResponse::Ok().json(ApiResponse {
            success: true,
            message: "Status retrieved".to_string(),
            data: Some(result),
            error: None,
        });
    }

    // Default to failed if we can't parse the response
    let result = VerificationResult {
        status: "failed".to_string(),
        guid: Some(req.guid.clone()),
        message: "Failed to determine verification status".to_string(),
    };

    HttpResponse::Ok().json(ApiResponse {
        success: true,
        message: "Status check completed".to_string(),
        data: Some(result),
        error: None,
    })
}

pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::resource("/verification/verify")
            .route(web::post().to(verify_contract))
    )
    .service(
        web::resource("/verification/status")
            .route(web::post().to(check_verification_status))
    );
}